### PR TITLE
Mysql incremental backups

### PR DESCRIPTION
**Pull request by a-shchupakov**
Added possibility to make incremental backups of MySQL-based servers

### CODE REVIEW

**Comment by a-shchupakov** [comment]
By the way, how do you grant permissions to run new integration test?

**Comment by g0djan** [comment]
chmod +x ?

File: internal/databases/mysql/backup_push_handler.go [hunk 1]
```diff
@@ -19,21 +94,80 @@ func HandleBackupPush(uploader *internal.Uploader, backupCmd *exec.Cmd) {
...
+		if err != nil {
+			tracelog.WarningLogger.Printf("Failed to read to_lsn. Future deltas cannot be done on top of the current backup: %v", err)
+		}
 	}
 
 	binlogEnd := getMySQLCurrentBinlogFile(db)
 	tracelog.DebugLogger.Println("Binlog end file", binlogEnd)
-	sentinel := StreamSentinelDto{BinLogStart: binlogStart, BinLogEnd: binlogEnd, StartLocalTime: timeStart}
 
+	sentinel := BackupSentinelDto{
```
Comment (mialinx) [inline internal/databases/mysql/backup_push_handler.go:120]: breaks backward compatibility

**Comment by ostinru** [comment]
Wow! It was an exciting patch :)

It seems that we should also fallback to full backup when `serer-uuid` have changed (I don't expect to have the same LSNs on all nodes of the cluster)

**Comment by x4m** [comment]
@ostinru +1
is there any chance to revivew this?

File: cmd/mysql/backup_push.go [hunk 2]
```diff
@@ -7,21 +7,31 @@ import (
...
+			mysql.HandleBackupPush(uploader, backupCmd, fullBackup)
+		},
+	}
+	fullBackup = false
+)
 
 func init() {
 	Cmd.AddCommand(backupPushCmd)
+
+	backupPushCmd.Flags().BoolVarP(&fullBackup, FullBackupFlag, FullBackupShorthand, false, "Make full backup-push")
```
Comment (mialinx) [inline cmd/mysql/backup_push.go:36]: This should be a default for backaward compatibility.
Incremental should be an option

File: internal/databases/mysql/backup_push_handler.go [hunk 3]
```diff
@@ -1,14 +1,89 @@
...
+}
+
+func createAndPushBackup(
+	uploader *internal.Uploader, backupCmd *exec.Cmd,
+	previousBackupSentinelDto internal.BackupSentinelDto,
+	incrementCount int,
+) {
+	extraLsnDir := path.Join("/tmp", "extra_lsn")
+	isXtrabackup := isXtrabackupCmd(backupCmd)
+	fromLSN, previousBackupName := previousBackupSentinelDto.BackupFinishLSN, previousBackupSentinelDto.IncrementFullName
```
Comment (mialinx) [inline internal/databases/mysql/backup_push_handler.go:75]: LSN in MySQL (innodb) has meaning only inside the single instance.
Backups may be taken from different instances (primary or one of secondary replicas).
So if current backup and previous are from different instances - we can rely on the LSN.

But we still can rely on LSN as long as we are taking backup from the same server

File: internal/databases/mysql/backup_fetch_handler.go [hunk 4]
```diff
@@ -1,16 +1,125 @@
...
+
+const (
+	XtrabackupApplyLogOnly   = "--apply-log-only"
+	XtrabackupIncrementalDir = "--incremental-dir"
 )
 
 func HandleBackupFetch(folder storage.Folder, backupName string, restoreCmd *exec.Cmd, prepareCmd *exec.Cmd) {
+	var err error
+	if isXtrabackupCmd(restoreCmd) {
+		tempDeltaDir := path.Join("/tmp", "delta")
```
Comment (mialinx) [inline internal/databases/mysql/backup_fetch_handler.go:24]: `/tmp`  is typically very limited in free space, database diff may take far more space
we need an option to choose directory for backup diffs

**Comment by a-shchupakov** [comment]
Feel free to catch up on this PR if you want. I'm afraid I don't remember a thing about it

**Comment by ostinru** [comment]
Closed as we merged another version of MySQL incremental backups (with some picks from this PR).
Thank you everyone for initial PR and Suggestions on how to improve it.

**Pull request closed without merging**